### PR TITLE
WIP: Strict validators

### PR DESCRIPTION
**Pull request by unconst**
This extension to the core validator ensures that the user sets the required number of weights on the chain, and also ensures that the max/min ratio is allowed. 

Changes:
-- new core validator compatible with new chain code
-- new template miner compatible with new chian code
-- minor text fixes in the cli
-- adds the query command into the cli

### CODE REVIEW

File: bittensor/_neuron/text/template_miner/__init__.py [hunk 1]
```diff
@@ -1,161 +1,702 @@
...
+            Arguments reflect an RPC backward request from another miner in the network. No response
+            needed for tokenized text inputs (uint64s have no gradient).
+
+            Args:
+                inputs_x ( :obj:`torch.Tensor`, `required`):
+                    torch inputs from previous forward call.
+                grads_dy ( :obj:`torch.Tensor`, `required`):
+                    torch grads of forward output.                    
+        """
+        if self.config.neuron.accumulate_remote_gradients:
```
Comment (Eugene-hu) [inline bittensor/_neuron/text/template_miner/__init__.py]: accumulate_remote_gradients is not part of the config

Comment (Eugene-hu) [reply bittensor/_neuron/text/template_miner/__init__.py]: I am also fairly certain that this old backward function does not work. It might be better to remove it and have the template miner only respond to forward requests

File: bittensor/_neuron/text/core_validator/__init__.py [hunk 2]
```diff
@@ -324,10 +330,11 @@ def forward(self, x: torch.tensor) -> torch.tensor:
 class nucleus( torch.nn.Module ):
     """ Nucleus class which holds the validator model.
     """
-    def __init__( self, config, device ):
+    def __init__( self, config, device, subtensor ):
         super(nucleus, self).__init__()
         self.config = config
         self.device = device
+        self.max_n = subtensor.max_n
```
Comment (isabella618033) [inline bittensor/_neuron/text/core_validator/__init__.py:341]: have we tested on nobunaga with a changing max_n value to make sure a changing size in the validator would not break anything?

File: bittensor/_neuron/text/template_miner/__init__.py [hunk 3]
```diff
@@ -1,161 +1,702 @@
...
+                '\n\t until_reset:', self.epoch % epochs_until_reset, '\n\t current_block:', current_block, '\n')
+        if self.config.using_wandb:
+            wandb.log( {    'batch_size': batch_size, 'sequence_length': sequence_length, 'n_topk_peer_weights': n_topk_peer_weights, 
+                            'max_allowed_ratio': max_allowed_ratio, 'blocks_per_epoch': blocks_per_epoch, 'epochs_until_reset': epochs_until_reset, 
+                }, step = current_block )
+
+        # === Reset Epochs with new params. ===
+        # Pulls new default validator training parameters and resets 
+        # the model and dataset for the following epoch.
+        if self.epoch % epochs_until_reset == 1:
```
Comment (Eugene-hu) [inline bittensor/_neuron/text/template_miner/__init__.py]: Should this be == 1? I think it is 0 on the core validator.

**Review by Eugene-hu — APPROVED** [review]
Looks fairly good to me; left you a few suggestions!

File: bittensor/_neuron/text/core_validator/__init__.py (hunk 2 above)
Comment (isabella618033) [reply bittensor/_neuron/text/core_validator/__init__.py:341]: also, do we have to change the n in metagraph as well? or it dosent matter

File: bittensor/_neuron/text/template_miner/__init__.py (hunk 1 above)
Comment (unconst) [reply bittensor/_neuron/text/template_miner/__init__.py]: Can we make this fix and allow gradients rather than remove
should simply require the config item

File: bittensor/_neuron/text/template_miner/__init__.py (hunk 3 above)
Comment (unconst) [reply bittensor/_neuron/text/template_miner/__init__.py]: yes
